src/python_routing_v02/troute/routing/fast_reach/simple_da.pyx

```diff
@@ -19,6 +20,63 @@ cpdef float simple_da_with_decay_py(
     )
 
 
+cdef float simple_da(
+    const float timestep,
+    const float routing_period,
+    const float decay_coeff,
+    const float gage_maxtimestep,
+    const float target_val,
+    const float model_val,
+    float lastobs_time,
+    float lastobs_val,
+) nogil:
+    """
+    wrapper function to compute all DA elements
+    """
+    cdef float replacement_val, nudge, da_weighted_shift, da_decay_minutes,
+    # cdef float lastobs_timestep, lastobs_value,
+
+    #printf("gages_size: %d\t", gages_size)
+    #printf("reach_has_gage[i]: %d\t", reach_has_gage[i])
+    #printf("num_reaches: %d\t", num_reaches)
+    #printf("i: %d\n", i)
+
+    # TODO: It is possible to remove the following branching logic if
+    # we just loop over the timesteps during DA and post-DA, if that
+    # is a major performance optimization. On the flip side, it would
+    # probably introduce unwanted code complexity.
+    if (timestep < gage_maxtimestep and not isnan(target_val)):
+        replacement_val = target_val
+        # add/update lastobs_timestep
+        lastobs_time = (timestep - 1) * routing_period
+        lastobs_val = target_val
+    else:
+        da_decay_minutes = ((timestep - 1) * routing_period - lastobs_time) / 60 # seconds to minutes
+        da_weighted_shift = obs_persist_shift(lastobs_val, model_val, da_decay_minutes, decay_coeff)
+        nudge = da_weighted_shift
+        # TODO: we need to export these values
+        replacement_val = simple_da_with_decay(lastobs_val, model_val, da_decay_minutes, decay_coeff)
+
+        # if gage_i == da_check_gage:
+        #     printf("gages_size: %d\t", gages_size)
+        #     printf("reach_has_gage[i]: %d\t", reach_has_gage[i])
+        #     printf("num_reaches: %d\t", num_reaches)
+        #     printf("i: %d\t", i)
+
+        #     printf("ts: %d\t", timestep)
+        #     printf("maxts: %d\t", gage_maxtimestep)
+        #     printf("a: %g\t", a)
+        #     printf("min: %g\t", da_decay_minutes)
+        #     printf("lo_ts: %d\t", lastobs_timestep[gage_i])
+        #     printf("ndg: %g\t", da_weighted_shift)
+        #     printf("lov: %g\t", lastobs_val)
+        #     printf("orig: %g\t", model_val)
+        #     printf("new: %g\t", replacement_val)
+        #     printf("\n")
+        return replacement_val
+        # return nudge, replacement_val
```

Comment:
@groutr The goal is to actually return these two values, not just the one. We could return the nudge only, then do the arithmetic (replacement_val = model_val + nudge) in the calling code.

Reply:
Resolved. See use of ctuple to declare function in simple_da.pyx:
``` py
cdef (float, float, float, float) simple_da(
...
```
... and the buffer to receive the results in mc_reach.pyx:
``` py
cdef (float, float, float, float) da_buf
...
    da_buf = simple_da(
        timestep,
...
```